File: src/code/cpp/graph/bellman_ford.cpp
#include <vector>
#include <utility>
#include <limits>

using namespace std;


vector<int> bellman_ford(int n, vector<pair<pair<int, int>, int>>& edges, int source) {
    vector<int> distances(n, numeric_limits<int>::max());
    distances[source] = 0;

    for (int i = 0; i < n - 1; ++i) {
        for (const auto& edge : edges) {
            int u = edge.first.first;
            int v = edge.first.second;
            int w = edge.second;

            if (distances[u] != numeric_limits<int>::max() && distances[u] + w < distances[v]) {
                distances[v] = distances[u] + w;
            }
        }
    }

    for (const auto& edge : edges) {
        int u = edge.first.first;
        int v = edge.first.second;
        int w = edge.second;

        if (distances[u] != numeric_limits<int>::max() && distances[u] + w < distances[v]) {
            return vector<int>();
        }
    }

    return distances;
}

File: src/code/ruby/graph/bellman_ford.rb
def bellman_ford(n, edges, source)
  distances         = Array.new(n, Float::INFINITY)
  distances[source] = 0

  (n - 1).times do
    edges.each do |u, v, w|
      if distances[u] != Float::INFINITY && distances[u] + w < distances[v]
        distances[v] = distances[u] + w
      end
    end
  end

  edges.each do |u, v, w|
    if distances[u] != Float::INFINITY && distances[u] + w < distances[v]
      return []
    end
  end

  distances
end
